File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.java
import java.util.*;
class Solution {
    public int solution(int sticker[]) {
        if (sticker.length == 1) return sticker[0];
        return Math.max(helper(Arrays.copyOfRange(sticker,0,sticker.length - 1)), helper(Arrays.copyOfRange(sticker, 1,sticker.length)));
    }
    public int helper(int[] sticker) {
        int n = sticker.length;
        
        if (n == 0) return 0; // 빈 배열이면 0 리턴
        if (n == 1) return sticker[0]; // 길이가 1이면 해당 값 리턴
        int[] dp = new int[n];
        dp[0] = sticker[0];
        dp[1] = Math.max(sticker[0], sticker[1]);
        //애매한게 두가지있음 .첫번쨰는 인덱스가 해당칸 포함인지
        //두번째는 dp[i]으 ㅣ값이 그 선택지를 고른 상황인지.
        //나는 경험에 의해 안다 굳이 두개의 상태를 관리안해도 되는것을
        //결국 for문에서는 이전상태골랐을때 상태랑 안골랐을떄+현재골랐을떄 할거라서.
        for (int i = 2; i < n; i++) {
            dp[i] = Math.max(dp[i-1], dp[i-2] + sticker[i]);
        }
        return dp[n-1];
    }
}

File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.py
def solution(sticker):
    if len(sticker) == 1:
        return sticker[0]
    def helper(arr):
        if len(arr) == 0:
            return 0
        if len(arr) == 1:
            return arr[0]
        dp = [0] * len(arr) #dp[i]는 i번째 스티커에서 최댓값을 추적
        dp[0] = arr[0]
        dp[1] = max(arr[0], arr[1])
        
        for i in range(2, len(dp)):
            dp[i] = max(dp[i-2] + arr[i], dp[i-1])
        return dp[len(dp) - 1]
    
    return max(helper(sticker[1:]), helper(sticker[:-1]))
    